### PR TITLE
feat(meetings): updated meeting logging

### PR DESCRIPTION
**Pull request by jonniespratley**
This PR adds logging on user device setup and changes. This is meant to give more diagnostic information when debugging media issues.

Fixes #[239692](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-239692)

### Audio Logging
Here is the output from audio logging. When `muteAudio`, `unmuteAudio` is invoked.

```
wx-js-sdk Meeting:index#handleAudioLogging
wx-js-sdk audio
wx-js-sdk deviceId= default
wx-js-sdk label= Default - Display Audio (05ac:1107)
wx-js-sdk noiseSuppression= true
wx-js-sdk echoCancellation= true
wx-js-sdk autoGainControl= true
wx-js-sdk Meeting:index#unmuteAudio --> unmuted audio successfully
```

### Video Logging
Here is the output for video logging. When `muteVideo`, `unmuteVideo` is invoked.

```
wx-js-sdk Meeting:index#handleVideoLogging
wx-js-sdk video
wx-js-sdk deviceId= 8304431374e734fd20a3a3b81deac5f54a6c3152892f4dcf3fb2e33a1837b2f8
wx-js-sdk aspectRatio= 1.7777777777777777
wx-js-sdk frameRate= 29.969999313354492 
wx-js-sdk width= 1280
wx-js-sdk height= 720
wx-js-sdk Meeting:index#muteVideo --> muted video successfully
```

### Device Logging
Here is the output from device logging. When `addMedia` is invoked.

```
wx-js-sdk devices
wx-js-sdk device - Default - Display Audio (05ac:1107)
wx-js-sdk deviceId= default
wx-js-sdk groupId= cdee3ba00d91c6c665acc3a816afa4d49a0c7f365fac6e5a99fc852445f5ce1c
wx-js-sdk kind= audioinput
wx-js-sdk device - Display Audio (05ac:1107)
wx-js-sdk deviceId= 1741989fce3b7d701452bbf979305fc87c27e58b0f33ee916fcf10bea1eae5cc
wx-js-sdk groupId= cdee3ba00d91c6c665acc3a816afa4d49a0c7f365fac6e5a99fc852445f5ce1c
wx-js-sdk kind= audioinput
```

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 1]
```diff
@@ -2189,15 +2189,18 @@ export default class Meeting extends StatelessWebexPlugin {
...
 
+
       this.mediaProperties.setMediaSettings('audio', {
         echoCancellation: settings.echoCancellation,
         noiseSuppression: settings.noiseSuppression
       });
 
       LoggerProxy.logger.log('Meeting:index#setLocalAudioTrack --> Audio settings.', JSON.stringify(this.mediaProperties.mediaSettings.audio));
       this.mediaProperties.setLocalAudioTrack(audioTrack);
+      this.audioTrack = audioTrack;
```
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2203]: The values are already present in this.mediaproperties.audioTrack we dont need to save it again

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 2]
```diff
@@ -2249,6 +2255,16 @@ export default class Meeting extends StatelessWebexPlugin {
     if (localStream) {
       const {audioTrack, videoTrack} = MeetingUtil.getTrack(localStream);
 
+      audioTrack.onmute = (e) => this.handleAudioLogging(e);
+      audioTrack.onunmute = (e) => this.handleAudioLogging(e);
+      videoTrack.onended = (e) => this.handleVideoLogging(e);
+      videoTrack.onmute = (e) => this.handleVideoLogging(e);
+
+      // TODO - did this trigger? any standard?
+      // audioTrack.addEventListener('unmute', (e) => this.handleAudioLogging(e));
```
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2264]: why do we have a todo here ?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 3]
```diff
@@ -4406,4 +4433,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+    await this.handleDeviceLogging();
+  }
+
+  /**
+   * Handle audio player logging, dumps details such as autoGainControl, echoCancellation
+   * volume and deviceId
+   * @returns {Promise} resolves empty promise
+   * @memberof Meetings
+   */
+  handleAudioLogging = async () => {
```
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:4455]: would move this logic to meeting utils as people can access this function

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 4]
```diff
@@ -3315,6 +3341,7 @@ export default class Meeting extends StatelessWebexPlugin {
             });
           }
         })
+        .then(this.handleMediaLogging)
```
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:3344]: We still need to add the devicechange listner for the devices

**Review by arun3528 — CHANGES_REQUESTED** [review]
Thanks for the PR , Nice job writing it , There are some minor comments 

1) Have we done some testing on mobile and safari browser 
2) Can you also past how does it look when the user does a mute and unmute and initial setup on the call

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 5]
```diff
@@ -102,20 +102,20 @@ const logRequest = (request, {header = '', success = '', failure = ''}) => {
...
 
 /**
-  * AudioVideo
-  * @typedef {Object} AudioVideo
-  * @property {Object} audio
-  * @property {String} audio.deviceId
-  * @property {Object} video
-  * @property {String} video.deviceId
-  */
+ * AudioVideo
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:105]: Is your editor auto adjusting everything? Can you make sure to only make changes related to the scope of work?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 6]
```diff
@@ -2189,15 +2189,18 @@ export default class Meeting extends StatelessWebexPlugin {
    */
   setLocalAudioTrack(audioTrack, emitEvent = true) {
     if (audioTrack) {
+      // TODO - Add logic for more settings
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2192]: What does this TODO mean?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 2 above)
Comment (lalli-flores) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2264]: Piggybacking, let's remove any commented out code

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 3 above)
Comment (lalli-flores) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:4455]: Utility file here https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/%40webex/plugin-meetings/src/meeting/util.js

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 7]
```diff
@@ -4406,4 +4433,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+        LoggerProxy.logger.group(`device - ${d.label}`);
+        LoggerProxy.logger.info('deviceId=', d.deviceId);
+        LoggerProxy.logger.info('groupId=', d.groupId);
+        LoggerProxy.logger.info('kind=', d.kind);
+        LoggerProxy.logger.groupEnd(`device - ${d.label}`);
+      });
+    }
+    LoggerProxy.logger.groupEnd('devices');
+
+    return true;
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:4522]: What does this line do?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 8]
```diff
@@ -4406,4 +4433,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+      LoggerProxy.logger.log('deviceId=', deviceId);
+      LoggerProxy.logger.log('aspectRatio=', aspectRatio);
+      LoggerProxy.logger.log('frameRate=', frameRate);
+      LoggerProxy.logger.log('width=', width);
+      LoggerProxy.logger.log('height=', height);
+      LoggerProxy.logger.groupEnd('video');
+    }
+
+
+    return this;
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:4498]: Why are you returning `this`?

**Review by lalli-flores — CHANGES_REQUESTED** [review]
Really good start. Can you make sure to only make changes related to your code changes? Seems like things are being aligned differently, etc. Can you revert those? Also, remove any commented out code

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 5 above)
Comment (jonniespratley) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:105]: I thought `npm` runs `eslint --fix`?  But I will revert and check.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 7 above)
Comment (jonniespratley) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:4522]: ends the console group and returns a value. I can update the logging to not use groups and fall into the other logging of json values.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 8 above)
Comment (jonniespratley) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:4498]: eslint was asking for a return value and usage of `this` in class member.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 4 above)
Comment (jonniespratley) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:3344]: Where should this be placed?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 6 above)
Comment (jonniespratley) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: ?

**Comment by lalli-flores** [comment]
I forgot to mention, on commits. A commit should typically map to a feature (e.g. new capability for a developer), so you can have many commits as you address feedback, but before merging you should always squash your commit to the number of features/fixes being introduced. In this case, it's one, so there should only be one commit.
The commit scope typically matches the name of the plugin, and the commit message should be something that someone can read in the changelog and understand right away what changed.
A suggested commit message:

```
feat(meetings): log media device data on audio/video changes
```

Alternatively, you can keep a single commit and rebase feedback.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 9]
```diff
@@ -679,4 +679,45 @@ MeetingUtil.startInternalStats = (meeting) => {
...
+    LoggerProxy.logger.log(LOG_HEADER, 'settings', JSON.stringify(settings));
+  }
+};
+
+MeetingUtil.handleDeviceLogging = (devices) => {
+  const LOG_HEADER = 'MeetingUtil#handleDeviceLogging -->';
+
+  LoggerProxy.logger.log(LOG_HEADER);
+
+  if (devices && devices.length) {
```
Comment (jdaly13) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:715]: can this be called without devices?  can you just default devices to be an array and not have to do this check

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 10]
```diff
@@ -679,4 +679,42 @@ MeetingUtil.startInternalStats = (meeting) => {
...
+  const LOG_HEADER = 'MeetingUtil#handleVideoLogging -->';
+
+  LoggerProxy.logger.log(LOG_HEADER);
+
+  if (videoTrack) {
+    const settings = videoTrack.getSettings();
+    const {deviceId} = settings;
+
+    LoggerProxy.logger.log(LOG_HEADER, `deviceId = ${deviceId}`);
+    LoggerProxy.logger.log(LOG_HEADER, 'settings', JSON.stringify(settings));
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:706]: For consistency
```suggestion
    LoggerProxy.logger.log(LOG_HEADER, 'settings =', JSON.stringify(settings));
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 11]
```diff
@@ -679,4 +679,42 @@ MeetingUtil.startInternalStats = (meeting) => {
...
+MeetingUtil.handleAudioLogging = (audioTrack) => {
+  const LOG_HEADER = 'MeetingUtil#handleAudioLogging -->';
+
+  LoggerProxy.logger.log(LOG_HEADER);
+
+  if (audioTrack) {
+    const settings = audioTrack.getSettings();
+    const {deviceId} = settings;
+
+    LoggerProxy.logger.log(LOG_HEADER, deviceId);
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:691]: Consistency with other functions
```suggestion
    LoggerProxy.logger.log(LOG_HEADER, `deviceId = ${deviceId}`);
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 12]
```diff
@@ -679,4 +679,42 @@ MeetingUtil.startInternalStats = (meeting) => {
...
+    LoggerProxy.logger.log(LOG_HEADER, `deviceId = ${deviceId}`);
+    LoggerProxy.logger.log(LOG_HEADER, 'settings', JSON.stringify(settings));
+  }
+};
+
+MeetingUtil.handleDeviceLogging = (devices = []) => {
+  const LOG_HEADER = 'MeetingUtil#handleDeviceLogging -->';
+
+  LoggerProxy.logger.log(LOG_HEADER);
+  devices.forEach((d) => {
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:714]:
```suggestion
  devices.forEach((device) => {
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 13]
```diff
@@ -679,4 +679,42 @@ MeetingUtil.startInternalStats = (meeting) => {
...
+  const LOG_HEADER = 'MeetingUtil#handleAudioLogging -->';
+
+  LoggerProxy.logger.log(LOG_HEADER);
+
+  if (audioTrack) {
+    const settings = audioTrack.getSettings();
+    const {deviceId} = settings;
+
+    LoggerProxy.logger.log(LOG_HEADER, deviceId);
+    LoggerProxy.logger.log(LOG_HEADER, 'settings', JSON.stringify(settings));
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:692]:
```suggestion
    LoggerProxy.logger.log(LOG_HEADER, 'settings =', JSON.stringify(settings));
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 14]
```diff
@@ -2206,6 +2206,14 @@ export default class Meeting extends StatelessWebexPlugin {
 
       LoggerProxy.logger.log('Meeting:index#setLocalAudioTrack --> Audio settings.', JSON.stringify(this.mediaProperties.mediaSettings.audio));
       this.mediaProperties.setLocalAudioTrack(audioTrack);
+
+      this.mediaProperties.audioTrack.onmute = (e) => {
+        console.warn('muted', e);
```
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:2211]: Is this meant to be here? or are they supposed to be logs?

**Review by lalli-flores — CHANGES_REQUESTED** [review]
Don't forget to add [unit tests for the functions you just introduced in the meeting utilities](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/%40webex/plugin-meetings/test/unit/spec/meeting/utils.js)

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 14 above)
Comment (jonniespratley) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: cleaned up

**Comment by jonniespratley** [comment]
Ok, im writing the unit tests.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 11 above)
Comment (jonniespratley) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:745]: Sounds good!

**Comment by lalli-flores** [comment]
Replaced by https://github.com/webex/webex-js-sdk/pull/2125

**Pull request closed without merging**